module-system/sov-modules-api/src/lib.rs

```diff
@@ -265,3 +275,44 @@ pub trait RpcRunner {
     type Context: Context;
     fn get_storage(&self) -> <Self::Context as Spec>::Storage;
 }
+
+/// Sorts ModuleInfo objects by their dependencies
+pub fn sort_by_dependencies<C: Context>(
+    modules: Vec<&dyn ModuleInfo<Context = C>>,
+) -> Result<Vec<&dyn ModuleInfo<Context = C>>, anyhow::Error> {
+
+    fn visit_module<'a, C: Context>(
+        module: &'a dyn ModuleInfo<Context = C>,
```

Comment:
What do you think about making a `ModuleVisitor` struct that holds all the required data and a `visit` method?  This way, we could remove the explicit lifetime `a` as it will be inferred from `&mut self` + it will give a bite more structure to the code (but it I a matter of taste).

Reply:
Would have to try it, I don't have enough intuition about Rust yet

Reply:
I tried to refactor it to something like:

```rust
struct ModuleVisitor<C: Context> {
    visited: HashSet<<C as Spec>::Address>,
    modules: Vec<Box<dyn ModuleInfo<Context = C>>>,
}

impl<C: Context> ModuleVisitor<C>{
    fn visit_module(
        &mut self, 
        module: &dyn ModuleInfo<Context = C>,
        sorted_modules: &mut std::vec::Vec<&dyn ModuleInfo<Context = C>>,
    ) -> Result<(), anyhow::Error> {
        let address = module.address();
        
        // if the module hasn't been visited yet, visit it and its dependencies
        if self.visited.insert(*address) {
            for dependency_address in module.dependencies() {
                let dependency_module = self.modules.iter().find(|module| module.address() == dependency_address)
                    .ok_or_else(|| anyhow::Error::msg(format!("Module not found: {:?}", dependency_address)))?
                    .as_ref();
                self.visit_module(dependency_module, sorted_modules)?;
            }
    
            sorted_modules.push(module);
        }
    
        Ok(())
    }
}
```
but failed to make it work.

Reply:
refactor in #570